Paper: Monaco: A Monte Carlo Library for Performing Uncertainty and Sensitivity Analyses (#711)
scipy-conference/scipy_proceedings: papers/scott_shambaugh/00_shambaugh.rst

:author: W\. Scott Shambaugh
:email: wsshambaugh@gmail.com
:bibliography: refs

-----------------------------------------------------------------------------------
Monaco: A Monte Carlo Framework for Performing Uncertainty and Sensitivity Analyses
-----------------------------------------------------------------------------------

.. class:: abstract

   This paper introduces *monaco*, a Python library for conducting Monte Carlo simulations of computational models, and performing uncertainty analysis (UA) and sensitivity analysis (SA) on the results. UA and SA are critical to effective and responsible use of models in science, engineering, and public policy, however their use is uncommon and hampered by lack of tools and need for advanced statistical knowledge. By providing a simple, general, and rigorous-by-default framework that wraps around existing models, *monaco* makes UA and SA easy and accessible to practitioners with a basic knowledge of statistics.

.. class:: keywords

   Monte Carlo, Modeling, Uncertainty Quantification, Uncertainty Analysis, Sensitivity Analysis, Ensemble Prediction, VARS, D-VARS

Introduction
============

> Computational models form the backbone of decision-making processes in science, engineering, and public policy. However, our increased reliance on these models stands in contrast to the difficulty in understanding them as we add increasing complexity to try and capture ever more of the fine details of real-world interactions. Practitioners will often take the results of their large, complex model as a point estimate, with no knowledge of how uncertain those results are. Multiple-scenario modeling (e.g. looking at a worst-case, most-likely, and best-case scenario) is an improvement, but what is needed is a complete global exploration of the input space. That gives insight into the overall distribution of results (UA) as well as the relative influence of the different input factors on the ouput variance (SA). This complete understanding is critical for effective and responsible use of models in any decision-making process.

Reviewer: While potentially backed up later, this paragraphs contains a lot of unsubstantiated claims about the field and common practices. Would it be possible to provide references right away?

Author: Added earlier reference.

Typically UA and SA are done using Monte Carlo simulations, for reasons explored in the following section. There are Monte Carlo frameworks available, however existing options are domain-specific, focused on narrow sub-problems (i.e. integration), tailored towards training neural nets, or require a deep statistical background to use. See :cite:`OLIVIER2020101204` and :cite:`razavi2021future` for an overview of the currently available Python tools for performing UA and SA respectively. For the domain expert who is not a statistician and wants to perform UA and SA on their existing models, there are no good options. *monaco* was written to address this gap.

.. figure:: monaco_logo.png
   :align: center
   :figclass: h
   :scale: 20 %

   The monaco project logo. :label:`monacologo`

> Why Monte Carlos?

Reviewer: This sounds odd, what are "Monte Carlo_s_"?
```suggestion
Motivation for Monte Carlo approach
```

Author: Changed to suggestion

To sample :math:`k` input factors where each factor :math:`x_{i \in [1, ..., k]}` is drawn from a unique probability distribution, it is a common practice to uniformly draw a random percentile between 0 and 1, :math:`p_i \in U[0, 1]`, and transform this to the target distribution through that distribution's inverse cumulative density function (CDF), :math:`x_i = F_i^{-1}(p_i)`. Thus the preceding result on the unit hypercube is easily generalized to an arbitrarily distributed input space.

When to Use and When to Avoid Monte Carlo Analysis
--------------------------------------------------

A complete exploration of a model's input space is necessary to fully characterize its behavior, and responsible use of computational models requires understanding the total range of predictions they imply. Policy papers have identified UA and SA as a critical good modeling practice :cite:`azzini2020uncertainty` :cite:`us2009guidance`. The short answer to the question of when to conduct UA and SA is, "always".

> There are some important considerations to keep in mind, however. With computational power making running large numbers of cases ever easier, a Monte Carlo analysis can result in highly statistically signficant conclusions. However these results are *conditional on the correctness of the underlying model and input distributions*. If the underlying model has not been throroughly validated, then any precise quantification of uncertainty and sensitivities will be washed out by the mismatch between the model and reality.

Reviewer: This and the next paragraph appear a bit "self-evident" and do not provide much substance to the paper in my opinion.

Author: Cut this down significantly.

This validation can be difficult - outputs are a function of the combined inputs and model, and cannot be used to validate either the inputs or the model on their own. Generally validation requires significant domain expertise to ensure a mechanistic model has a solid theoretical foundation, and to diagnose errors in its implementation. 

> This is not to say that UA and SA should not be conducted early in the model development process - obtaining the range of plausible output uncertainties is a critical step in input and model validation. Test data cannot be well compared against a single point estimate of a model's output, and it is necessary to have the full distribution of output values to compare test data against. Once a Monte Carlo analysis has generated these distributions, hypothesis testing or probablistic prediction measures like loss scores can be used to anchor the outputs against real-life test data.

Reviewer: This and the preceeding paragraphs are a bit too much "waffle". The paper would benefit from staying more focused on the topic at hand, the particular software that is presented, and stay clear  of making an attempt at providing a holistic motivation for the underlying methodology.

Author: Cut this down significantly

While Monte Carlo analysis is not strictly necessary for linear models, it is often an easier and conceptually simpler way to compute the propogation of uncertainties and sensitivities through a model than using linear methods. And by making nonlinear models easier to examine, there should be less of a need to make linearity assumptions about a system in the first place.

One inherent pitfall of Monte Carlo approaches is that rare events may be undersampled. For example, NASA uses Monte Carlo simulations extensively during launch vehicle design to predict the rocket trajectory and performance. :cite:`hanson2010applying` However, they must prove robustness to anomalous or stressing scenarios which may occur only one or two times in a run of thousands of cases, which is not enough to draw conclusions from. In instances such as this, rare event scenarios should be investigated directly.

> Note that *monaco*'s computational and storage overhead in creating easily-iterrogatable objects for each variable, value, and case makes it an inefficient choice for computationally simple applications with high :math:`n`, such as Monte Carlo integration. It is best suited for models with moderate to high computational cost. 

Reviewer: So, for what range of _n_ is it suitable? What would be considered moderate to high computational costs? What is the computational complexity? Are there any benchmarks?

Author: Added some numbers on storage and computational costs, this is going to be very model dependent however

Why Use Basic Monte Carlo over Bayesian Methods?
------------------------------------------------

> *monaco* purposefully eschews the greater computational efficiency and built-in input parameter estimation of Bayesian methods such as Markov Chain Monte Carlo in favor of basic sampling. For its target audience of scientists, engineers, and policy analysts with mechanistic models, this is useful for many reasons:

Reviewer: I think it is borderline absurd to claim that basic techniques such as MCMC would be too complex for scientists, engineers, and policy analysists to understand.

Already considering the next paragraph, it is absolutely possible for software with well-designed API to be suitable to both novice and advanced users. Using layered APIs and appropriate defaults, complexity can be effectively hidden, while still empowering the expert user to embrace them when needed.

Author: Cut this down significantly, will address your point in separate comment

> * `monaco` favors conceptual simplicity for greater accessibility. The goal is to be usable by someone at the level of knowing what a Uniform and Normal distribution are, and not require an in-depth data science background in inference. For one example of a barrier to entry that would not be covered in an undergraduate education, Bayesian methods generate inherently correlated sample points, which invalidates the independence assumption of many basic statistical methods one might want to apply to outputs. 

Reviewer: I would have expected that a tool that claims to be a "Monte Carlo _Framework_" would at least provide the option for something other than basic random sampling.

Author: Didn't realize the extent to which framework was a term of art! Changed to "library" everywhere

> * Many modeling domains are not data-rich, especially in contrast to the dimensionality :math:`k` of the model. This precludes using standard Bayesian approaches to obtain results with any useful level of confidence. Consider NASA's use of Monte Carlo in predicting a rocket's flight :cite:`hanson2010applying` - they might have high levels of confidence that the hundreds of parameters describing their rocket are properly bound, and that their model correctly implements the physics of flight, but before their first launch they won't have any flight data to anchor against at all. UA and SA are still critical tools in these situations.

Reviewer: I feel like these broad claims about various modeling domains are overly reductive.

Reviewer 2: @scottshambaugh  I have gone through the paper and the paper seems to be a thorough effort at developing a comprehensive library to support monte carlo simulations. The paper elaborates very well the working of the library and focusses on the key objects via examples. At a macro level, I can say that the paper fulfils all the key aspects that it should cover. At micro level analysis with separate sections under consideration,I would strongly suggest, to add a few case studies (one or two in no.) application areas where monte carlo simulations are used. For example, it is mentioned that NASA utilises monte carlo simulations for rocket trajectory predictions,it would be better if this or another case study is implemented utilising this library. This will enhance the purpose of library and motivate the readers to experiment with it. Kindly incorporate.

Author: csadorf, reworked this expensively and removed claims about modeling domains.
@shubhamsharma1609, I've added two case studies to the end of the paper

Workflow
--------

UA and SA of any model follows a common workflow. Probability distributions for the model inputs are defined, and randomly sampled values for a large number of cases are fed to the model. The outputs from each case are collected and the full set of inputs and outputs can be analyzed. Typically UA is performed by generating histograms, scatter plots, and summary statistics for the output variables, and SA is performed by looking at the effect of input on output variables through scatter plots and calculating sensitivity indices. These results can then be compared to real-world test data to validate the model or inform revisions to the model and input variables. See Fig. :ref:`figanalysisprocess`.

Note that *monaco* does not currently have tools for model or parameter validation, and closing that part of the workflow loop is left up to the user.

.. figure:: analysis_process.png
   :align: center
   :figclass: h

   Monte Carlo workflow for understanding the full behavior of a computational model, inspired by :cite:`saltelli2019so`. :label:`figanalysisprocess`

*monaco* Structure
==================

Overall Structure
-----------------

> Broadly, each input factor and model output is a *variable* that can be thought of as lists (rows) containing the full range of randomized *values*. *Cases* are slices (columns) that take the *i*'th input and output value for each variable, and represent a single run of the model. Each case is run on its own in parallel, and the output values are collected into output variables. Fig. :ref:`figarchitecture` shows a visual representation of this.

Reviewer: What parallelization technique is used?

It is clarified later that the software uses dask for parallel processing, might be worthwhile to at least mention here.

Author: I just took it out of this section

Data Flow
---------

A summary of the process and data flow:

1) Instantiate a `Sim` object.
2) Add input variables to the sim with specified probability distributions.
3) Run the simulation. This executes the following:    

 a) Random percentiles are drawn `ndraws` times for each of the input variables.
 b) These percentiles are transformed into random values via the inverse CDF of the target probability distribution.
 c) If nonnumeric inputs are desired, the random numbers are converted to objects via a `nummap` dict.
 d) `Case` objects are created and populated with the input values for each case.
 e) Each case is run by structuring the inputs values with the `preprocess` function, passing them to the `run` function, and collecting the output values with the `postprocess` function.
 f) The output values are collected into output variables and saved back to the sim. If the values are nonnumeric, a `valmap` dict assigning numbers to each unique value is automatically generated.

4) Calculate statistics & sensitivities for input & output variables.
5) Plot variables, their statistics, and sensitivities.

Technical Features
==================

Sampling Methods
----------------

> Random sampling of the percentiles for each variable can be done using scipy's pseudo-random number generator (PRNG), or with any of the low-discrepancy methods in `scip.stats.qmc` Quasi-Monte Carlo module. In general, the `'sobol_random'` method that generates Sobol sequences with Owen scrambling :cite:`sobol1967distribution` :cite:`owen2020dropping` is recommended in nearly all cases as a well-performing quasi-random sequence with the best known convergence, balanced integration properties as long as the number of cases is a power of 2, and a fairly flat frequency spectra :cite:`perrier2018sequences`. This is set as default. In cases where computing sample points takes a  prohibitively long amount of time, users may fall back to `'random'` sampling directly from the PRNG at the cost of less even distribution of points in the input space. See Fig. :ref:`figsampling` for a visual comparison.

Reviewer: Using Sobol sequences, wouldn't this classify the algorithm as Quasi-Monte Carlo?

Further, looking at the uniform distribution and the frequency spectrum plot, there are some very clear artifacts that I would expect to lead to biased sampling with respect to specific intervals. Maybe you can comment on the effect/non-effect of those correlations on the UA/SA results?

Author: Yes, made the QMC distinction more explicit. Added a note on the spectra and some more justification for the default being scrambled Sobol sampling

    sim.outvars['Sum'].plotSensitivities()

.. figure:: example.png
   :align: center
   :figclass: h

   Output from the example code which calculates the sum of two random dice rolls. The top plot shows a histogram of the 2-dice sum with the mean and 5 - 95th percentiles marked, the middle plot shows the randomness over the set of rolls, and the bottom plot shows that each of the dice contributes 50% to the variance of the sum. :label:`figexample`

Conclusion
==========

This paper has introduced the ideas underlying Monte Carlo analysis and discussed when it is appropriate to use for conducting UA and SA. It has shown how *monaco* implements a rigorous, parallel Monte Carlo framework, and how to use it through a simple example. This library is geared towards scientists, engineers, and policy analysts that have a computational model in their domain of expertise, enough statistical knowledge to define a probability distribution, and a desire to ensure their model will make accurate predictions of reality. The author hopes this tool will help contribute to easier and more widespread use of UA and SA in improved descision-making.

Further Information
===================

> *monaco* is available on PyPI, has API documentation at https://monaco.rtfd.io/, and is hosted on github at https://github.com/scottshambaugh/monaco/. Please see the "examples" directory in the github source for many more Monte Carlo implementation examples across a range of domains such as physics simulation, election prediction, financial modeling, pandemic spread, and integration.

Reviewer: I'd suggest to mention that is available on PyPI under the same name (`monaco`). While this may seem obvious, this just removes any doubt that this is the right package.

Author: Good idea, done

* The same sample points can be used for UA and SA. Generally, Bayesian methods such as Markov Chain Monte Carlo provide much faster convergence on UA quantities of interest, but their undersampling of regions that do not contribute to the desired quantities is inadequate for SA and complete exploration of the input space. The author's experience aligns with :cite:`saltelli2019so` in that there is great practical benefit in broad sampling without pidgeonholing one's purview to particular posteriors, through uncovering bugs and edge cases in regions of input space that were not being previously considered.
* It can be applied to domains that are not data-rich. See for example NASA's use of Monte Carlo simulations during rocket design prior to collecting test flight data :cite:`hanson2010applying`.

However, basic Monte Carlo sampling is subject to the classical drawbacks of the method such as poor sampling of rare events and the slow :math:`\sigma / \sqrt{n}` convergence on quantities of interest. If the outputs and regions of interest are firmly known at the outset, then other sampling methods will be more efficient :cite:`kroese2013handbook`.

> Additionally, given that any conclusions are conditional on the correctness of the underlying model and input parameters, the task of validation is critical to confidence in the UA and SA results. However, this is currently out of scope for the library and must be performed with other tools. In a data-poor domain, hypothesis testing or probablistic prediction measures like loss scores can be used to anchor the outputs against a small number of real-life test data. More generally, the "inverse problem" of model and parameter validation is a deep field unto itself and :cite:`national2012assessing` and :cite:`shiffrin2008survey` are recommended as overviews of some methods. If *monaco*'s scope is too limited for the reader's needs, the author recommends `UQpy` :cite:`OLIVIER2020101204` for UA and SA, and `PyMC` :cite:`salvatier2016probabilistic` or `Stan` :cite:`carpenter2017stan` as good general-purpose probablistic programming Python libraries.

Reviewer 3: typo: probablistic > probabilistic

Workflow
--------

UA and SA of any model follows a common workflow. Probability distributions for the model inputs are defined, and randomly sampled values for a large number of cases are fed to the model. The outputs from each case are collected and the full set of inputs and outputs can be analyzed. Typically UA is performed by generating histograms, scatter plots, and summary statistics for the output variables, and SA is performed by looking at the effect of input on output variables through scatter plots, performing regressions, and calculating sensitivity indices. These results can then be compared to real-world test data to validate the model or inform revisions to the model and input variables. See Fig. :ref:`figanalysisprocess`.

> Note again that *monaco* does not currently have tools for model or input parameter validation, and closing that part of the workflow loop is left up to the user. 

Reviewer: ```suggestion
Model or input parameter validation does not fall into *monaco*'s scope; closing that part of the workflow loop is left up to the user.
```

Author: Reworded

    simoutput = (simoutput1, simoutput2, ...)
                  = run(*siminput) 
                  = run(siminput1, siminput2, ...)
    _ = postprocess(case, *simoutput)
      = postprocess(case, simoutput1, simoutput2, ...)

These three functions must be passed to the simulation in a dict with keys `'run'`, `'preprocess'`, and `'postprocess'`. See the example code at the end of the paper for a simple worked example.

Examining Results
-----------------

After running, users should generally do all of the following UA and SA tasks to get a full picture of the behavior of their computational model.

* Plot the results (UA & SA).

* Calculate statistics for input or output variables (UA).

* Calculate sensitivity indices to rank importance of the input variables on variance of the output variables (SA).

* Investigate specific cases with outlier or puzzling results.

* Save the results to file or pass them to other programs. 

Data Flow
---------

A summary of the process and data flow:

1) Instantiate a `Sim` object.
2) Add input variables to the sim with specified probability distributions.
3) Run the simulation. This executes the following:    

>  a) Random percentiles :math:`p_i \in U[0,1]` are drawn `ndraws` times for each of the input variables.

Reviewer: This is rendered oddly.

Author: I have not been able to figure out how to get it to indent normally. @mepa?

Author: Opened an issue here: https://github.com/scipy-conference/scipy_proceedings/issues/741

Reviewer 4: The rst indentation is three spaces, not one:

<img width="421" alt="Screen Shot 2022-06-28 at 8 01 37 PM" src="https://user-images.githubusercontent.com/6868396/176328936-96a22b75-d292-4b6f-95ea-f0e2cfca8ed2.png">

More examples here: https://sublime-and-sphinx-guide.readthedocs.io/en/latest/lists.html

Author: Thank you!

 b) These percentiles are transformed into random values via the inverse cumulative density function of the target probability distribution.
 c) If nonnumeric inputs are desired, the random numbers are converted to objects via a `nummap` dict.
 d) `Case` objects are created and populated with the input values for each case.
 e) Each case is run by structuring the inputs values with the `preprocess` function, passing them to the `run` function, and collecting the output values with the `postprocess` function.
 f) The output values are collected into output variables and saved back to the sim. If the values are nonnumeric, a `valmap` dict assigning numbers to each unique value is automatically generated.

4) Calculate statistics & sensitivities for input & output variables.
5) Plot variables, their statistics, and sensitivities.

Incorporating into Existing Workflows 
-------------------------------------

> If the user want to use existing workflows for generating, running, post-processing, or examining results, any combination of *monaco*'s major steps can be replaced by external tooling by saving and loading input and output variables to file. For example, *monaco* can be used only for its parallel processing backend by importing existing randomly drawn input variables, running the simulation, and exporting the output variables for outside analysis. Or, it can be used only for its plotting and analysis capabilities by feeding it inputs and outputs generated elsewhere.

Reviewer: wants*

Author: Fixed

Resource Usage
--------------

> Note that *monaco*'s computational and storage overhead in creating easily-iterrogatable objects for each variable, value, and case makes it an inefficient choice for computationally simple applications with high :math:`n`, such as Monte Carlo integration. If the preprocessed sim input and raw output for each case (which for some models may dominate storage) is not retained, then the storage bottleneck will be the creation of a `Val` object for each case's input and output values with minimum size 0.5 kB. The maximum :math:`n` will be driven by the size of the RAM on the host machine being capable of holding at least :math:`0.5 * n(k_{in} + k_{out})` kB. On the computational bottleneck side, *monaco* is best suited for models where the model runtime dominates the random variate generation and the few hundred microseconds of `dask.delayed` task switching time. 

Reviewer 3: Is "iterrogatable" is a word? Maybe a typo?

.. figure:: sampling.png
   :align: center
   :figclass: h
   :scale: 25 %

   256 uniform and normal samples along with the 2D frequency spectra for PRNG random sampling (top), Sobol sampling (middle), and scrambled Sobol sampling (bottom, default). :label:`figsampling`

Order Statistics, or, How Many Cases to Run?
--------------------------------------------

> How many Monte Carlo cases should one run? One answer would be to choose :math:`n \geq 2^k` with a sampling method that implements a (t,m,s) digital net (such as a Sobol or Halton sequence), which guarentees that there will be at least one sample point in every hyperoctant of the input space :cite:`joe2008constructing`. This should be considered a lower bound for SA, with the number of cases run being some integer multiple of :math:`2^k`.

Reviewer 3: typo: guarentees -> guarantees

A rigorous way of choosing the number of cases is to first choose a statistical interval (e.g. a confidence interval for a percentile, or a tolerance interval to contain a percent of the population), and then use order statistics to calculate the minimum :math:`n` required to obtain that result at a desired confidence level. This approach is independent of :math:`k`, making UA of high-dimensional models tractable. *monaco* implements order statistics routines for calculating these statistical intervals with a distribution-free approach that makes no assumptions about the normality or other shape characteristics of the output distribution. See Chaper 5 of :cite:`hahn1991statistical`.

> A more qualitative UA method would simply to choose a reasonably high :math:`n` (say, :math:`n=2^{10}`), manually examine the results to ensure high-interest areas are not being undersampled, and rely on bootstrapping of the desired variable statistics to obtain the required confidence levels. 

Reviewer: would simply choose*

Author: Reworded
